der/src/document.rs

```diff
@@ -1,67 +1,102 @@
+use {
+    crate::pem::{self, PemLabel},
+    alloc::string::String,
+};
 
 #[cfg(feature = "std")]
 use std::{fs, path::Path};
 
+/// Marker trait which identifies documents which contain sensitive
+/// information, such as private keys.
+pub trait Sensitivity {
+    /// Does this type contain potentially sensitive data?
+    ///
+    /// This enables hardened file permissions when persisting data to disk.
+    const SENSITIVE: bool;
+}
+
 /// ASN.1 DER-encoded document.
 ///
-/// This trait is intended to impl on types which contain an ASN.1 DER-encoded
-/// document which is guaranteed to encode as the associated `Message` type.
+/// This type wraps an encoded ASN.1 DER message which is guaranteed to
+/// infallibly decode as type `T`.
 ///
 /// It implements common functionality related to encoding/decoding such
 /// documents, such as PEM encapsulation as well as reading/writing documents
 /// from/to the filesystem.
 #[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
-pub trait Document<'a>: AsRef<[u8]> + Sized + TryFrom<Vec<u8>, Error = Error> {
-    /// ASN.1 message type this document decodes to.
-    type Message: Decodable<'a> + Encodable + Sized;
+pub struct Document<T: Sensitivity> {
+    /// ASN.1 DER-encoded document guaranteed to decode to `T` infallibly.
+    der_bytes: Vec<u8>,
 
-    /// Does this type contain potentially sensitive data?
-    ///
-    /// This enables hardened file permissions when persisting data to disk.
-    const SENSITIVE: bool;
+    /// Rust type corresponding to the ASN.1 DER message the bytes can be
+    /// infallibly deserialized as.
+    msg_type: PhantomData<T>,
+}
 
-    /// Borrow the inner serialized bytes of this document.
-    fn as_der(&self) -> &[u8] {
-        self.as_ref()
-    }
+impl<T> Document<T> where T: Sensitivity {}
 
-    /// Return an allocated ASN.1 DER serialization as a boxed slice.
-    fn to_der(&self) -> Box<[u8]> {
-        self.as_ref().to_vec().into_boxed_slice()
+impl<T> Document<T>
+where
+    T: Sensitivity,
+{
+    /// Borrow the inner serialized bytes of this document.
+    pub fn as_der(&self) -> &[u8] {
+        self.der_bytes.as_slice()
     }
 
     /// Decode this document as ASN.1 DER.
-    fn decode(&'a self) -> Self::Message {
-        Self::Message::from_der(self.as_ref()).expect("ASN.1 DER document malformed")
+    pub fn decode<'a>(&'a self) -> T
+    where
+        T: Decodable<'a> + Sized,
+    {
+        self.try_decode().expect("ASN.1 DER document malformed")
     }
 
     /// Create a new document from the provided ASN.1 DER bytes.
-    fn from_der(bytes: &[u8]) -> Result<Self> {
-        bytes.to_vec().try_into()
+    pub fn from_der(bytes: impl Into<Vec<u8>>) -> Result<Self>
+    where
+        T: for<'a> Decodable<'a> + Sized,
+    {
+        let doc = Self {
+            der_bytes: bytes.into(),
+            msg_type: PhantomData,
+        };
+
+        // Ensure document parses successfully
+        doc.try_decode()?;
+        Ok(doc)
+    }
```

Comment:
I don't think we should try decoding the der. This has performance and security implications. I think the `Document` type should only have the invariant that it contains bytes successfully decoded from a PEM file. The important bit is that `T` is linked via the type system to the PEM document "name" and that we know the sensitivity of the bytes.